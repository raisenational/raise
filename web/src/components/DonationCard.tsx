import TimeAgo from 'react-timeago';
import classNames from 'classnames';
import { format } from '@raise/shared';

type Props = {
  donorName?: string,
  createdAt: number | string,
  giftAid?: boolean,
  comment?: string | null,
  currency?: 'gbp' | 'usd',
  donationAmount?: number,
  matchFundingAmount?: number,
  recurringAmount?: number | null,
  recurrenceFrequency?: ('WEEKLY' | 'MONTHLY') | null,
  className?: string,
  loading?: boolean,
};

const DonationCard: React.FC<Props> = ({
  donorName,
  createdAt,
  giftAid,
  comment,
  currency,
  donationAmount,
  matchFundingAmount,
  recurringAmount,
  recurrenceFrequency,
  className,
  loading,
}) => {
  const title = `${donorName || 'Someone'} donated${donationAmount !== undefined ? ` ${format.amountShort(currency, donationAmount)}` : ''}`;
  const isRecurring = recurringAmount !== undefined && recurringAmount !== null && recurrenceFrequency !== undefined && recurrenceFrequency !== null;
  const recurringText = isRecurring ? `giving ${format.amountShort(currency, recurringAmount)} ${recurrenceFrequency.toLowerCase()}` : undefined;
  const matchFundingText = (matchFundingAmount !== undefined && matchFundingAmount > 0) ? `+${format.amountShort(currency, matchFundingAmount)} matched` : undefined;
  const giftAidText = (giftAid && donationAmount !== undefined) ? `+${format.amountShort(currency, Math.floor(donationAmount * 0.25))} gift-aided` : undefined;
  const extraAmountText = (matchFundingText || giftAidText) ? `(${[matchFundingText, giftAidText].filter((x) => x).join(', ')})` : undefined;

  return (
    <div className={classNames('p-4 rounded flex flex-col shadow-raise', className)}>
      <p className="fade-in word-wrap"><span className={classNames({ skeleton: loading })}>{title}</span></p>
      {recurringText && <p className={classNames('fade-in', { skeleton: loading, 'text-base opacity-80': !loading })}>{recurringText}</p>}
      {extraAmountText && <p className="text-base fade-in"><span className={classNames({ skeleton: loading, 'opacity-80': !loading })}>{extraAmountText}</span></p>}
      {comment && <p className="text-base mt-2 word-wrap fade-in"><span className={classNames({ skeleton: loading })}>{comment}</span></p>}
      <p className="text-base mt-auto pt-2 text-right fade-in"><span className={classNames({ skeleton: loading, 'opacity-80': !loading })}>{typeof createdAt === 'string' ? createdAt : <TimeAgo date={createdAt * 1000} />}</span></p>
    </div>
  );
};

export default DonationCard;
