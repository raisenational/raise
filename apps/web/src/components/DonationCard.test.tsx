import { render } from '@testing-library/react';
import DonationCard from './DonationCard';

beforeAll(() => {
  Date.now = jest.fn(() => 1635699600000);
});

test.each([
  ['a date', <DonationCard createdAt={1635683696} />],
  ['a date string', <DonationCard createdAt="5 minutes ago" />],
  ['a name and date', <DonationCard donorName="Namey McNameface" createdAt={1635683696} />],
  ['amounts (match funding and gift-aid) and a date', <DonationCard currency="gbp" donationAmount={1234} matchFundingAmount={1234} giftAid createdAt={1635683696} />],
  ['amounts (match funding and no gift-aid) and a date', <DonationCard currency="gbp" donationAmount={1234} matchFundingAmount={1234} giftAid={false} createdAt={1635683696} />],
  ['amounts (no match funding and gift-aid) and a date', <DonationCard currency="gbp" donationAmount={1234} matchFundingAmount={0} giftAid createdAt={1635683696} />],
  ['amounts (no match funding and no gift-aid) and a date', <DonationCard currency="gbp" donationAmount={1234} matchFundingAmount={0} giftAid={false} createdAt={1635683696} />],
  ['amounts (whole numbers) and a date', <DonationCard currency="gbp" donationAmount={200_00} matchFundingAmount={200_00} giftAid createdAt={1635683696} />],
  ['a recurring amount and a date', <DonationCard currency="gbp" donationAmount={45_00} matchFundingAmount={45_00} giftAid recurringAmount={15_00} recurrenceFrequency="WEEKLY" createdAt={1635683696} />],
  ['a name and date', <DonationCard donorName="Namey McNameface" createdAt={1635683696} />],
  ['a custom class name', <DonationCard className="bg-raise-red" createdAt={1635683696} />],
  ['a loading state', <DonationCard className="bg-raise-red" donorName="Namey McNameface" currency="gbp" donationAmount={45_00} matchFundingAmount={45_00} giftAid recurringAmount={15_00} recurrenceFrequency="WEEKLY" createdAt={1635683696} />],
  ['a usd donation', <DonationCard className="bg-raise-red" donorName="Namey McNameface" currency="usd" donationAmount={45_00} matchFundingAmount={45_00} recurringAmount={15_00} recurrenceFrequency="WEEKLY" createdAt={1635683696} />],
])('renders with %s', (_, element) => {
  const { container } = render(element);
  expect(container).toMatchSnapshot();
});
