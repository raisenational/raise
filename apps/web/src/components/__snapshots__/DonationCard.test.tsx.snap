// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with a custom class name 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise bg-raise-red"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a date string 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        5 minutes ago
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a loading state 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise bg-raise-red"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Namey McNameface donated £45
      </span>
    </p>
    <p
      class="fade-in text-base opacity-80"
    >
      giving £15 weekly
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+£45 matched, +£11.25 gift-aided)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a name and date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Namey McNameface donated
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a name and date 2`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Namey McNameface donated
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a recurring amount and a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated £45
      </span>
    </p>
    <p
      class="fade-in text-base opacity-80"
    >
      giving £15 weekly
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+£45 matched, +£11.25 gift-aided)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with a usd donation 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise bg-raise-red"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Namey McNameface donated $45
      </span>
    </p>
    <p
      class="fade-in text-base opacity-80"
    >
      giving $15 weekly
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+$45 matched)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with amounts (match funding and gift-aid) and a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated £12.34
      </span>
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+£12.34 matched, +£3.08 gift-aided)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with amounts (match funding and no gift-aid) and a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated £12.34
      </span>
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+£12.34 matched)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with amounts (no match funding and gift-aid) and a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated £12.34
      </span>
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+£3.08 gift-aided)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with amounts (no match funding and no gift-aid) and a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated £12.34
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;

exports[`renders with amounts (whole numbers) and a date 1`] = `
<div>
  <div
    class="p-4 rounded flex flex-col shadow-raise"
  >
    <p
      class="fade-in word-wrap"
    >
      <span
        class=""
      >
        Someone donated £200
      </span>
    </p>
    <p
      class="text-base fade-in"
    >
      <span
        class="opacity-80"
      >
        (+£200 matched, +£50 gift-aided)
      </span>
    </p>
    <p
      class="text-base mt-auto pt-2 text-right fade-in"
    >
      <span
        class="opacity-80"
      >
        <time
          datetime="2021-10-31T12:34:56.000Z"
          title="2021-10-31 12:34"
        >
          4 hours ago
        </time>
      </span>
    </p>
  </div>
</div>
`;
